### ISSUE TITLE
Can the authors provide an explanation on why the KMO matcher isn't reflected in the code?

### ISSUE THREAD

**Issue by YuchenWg** [dk-liang/CLTR#31]
The matcher.py file in the repository implements the basic L1-based Hungarian matcher and does not reflect the _KMO-based Hungarian bipartite matching_ claimed in the paper.

Looking through the closed issues, it seems like others have already pointed this out but the authors seem to have closed the issues without providing an explanation. 

Given that the KMO matcher is really the crux of the paper, without which the model become merely a clone of the conditional DETR adapted to output points, this issue will inevitably be raised again in the future. 

To avoid this issue being repeated raised in the repository, can the authors provide an explanation? Is it out of IP concerns? Is it still being implemented?

**Comment by tejasri19** [comment]
Any update on the KMO-matcher ? @dk-liang

**Comment by tejasri19** [comment]
@YuchenWg Any thoughts on the below code for KMO-matcher:

`class KMOHungarianMatcher_Crowd(nn.Module):
    """This class computes an assignment between the targets and the predictions of the network using the KMO-based matching."""
    def __init__(self, cost_class: float = 1, cost_point: float = 1, k_nearest_neighbors: int = 3):
        super().__init__()
        self.cost_class = cost_class
        self.cost_point = cost_point
        self.k_nearest_neighbors = k_nearest_neighbors
        assert cost_class != 0 or cost_point != 0, "all costs can't be 0"

    @torch.no_grad()
    def forward(self, outputs, targets):
        bs, num_queries = outputs["pred_logits"].shape[:2]
        device = outputs["pred_logits"].device

        # Handle empty batch
        if bs == 0:
            return []

        indices = []
        for b in range(bs):
            # Get predictions for this batch
            out_prob = outputs["pred_logits"][b]  # [num_queries, num_classes]
            out_points = outputs["pred_points"][b]  # [num_queries, 2]

            # Get targets for this batch
            tgt_ids = targets[b]["labels"]
            tgt_points = targets[b]["point"]

            # Handle empty targets
            if tgt_points.shape[0] == 0 or tgt_ids.shape[0] == 0:
                indices.append((
                    torch.empty(0, dtype=torch.int64, device=device),
                    torch.empty(0, dtype=torch.int64, device=device)
                ))
                continue

            # Handle empty predictions
            if num_queries == 0:
                indices.append((
                    torch.empty(0, dtype=torch.int64, device=device),
                    torch.empty(0, dtype=torch.int64, device=device)
                ))
                continue

            # Classification cost
            foreground_class_index = 0
            C_hat_j = out_prob[:, foreground_class_index]  # [num_queries]

            # Point distance cost
            cost_point = torch.cdist(out_points, tgt_points, p=1)  # [num_queries, num_tgt_points]

            # Context cost for targets
            tgt_dists = torch.cdist(tgt_points, tgt_points, p=1)  # [num_target_points, num_target_points]
            tgt_dists.fill_diagonal_(float('inf'))
            
            k = min(self.k_nearest_neighbors, max(1, tgt_points.shape[0] - 1))
            if k > 0 and tgt_points.shape[0] > 1:
                nearest_dists_tgt, _ = tgt_dists.topk(k, largest=False, dim=1)
                y_k_i = nearest_dists_tgt.mean(dim=1)
            else:
                y_k_i = torch.zeros(tgt_points.shape[0], device=device)

            # Context cost for predictions
            pred_dists = torch.cdist(out_points, out_points, p=1)
            pred_dists.fill_diagonal_(float('inf'))
            
            k_pred = min(self.k_nearest_neighbors, max(1, num_queries - 1))
            if k_pred > 0 and num_queries > 1:
                nearest_dists_pred, _ = pred_dists.topk(k_pred, largest=False, dim=1)
                y_k_j_hat = nearest_dists_pred.mean(dim=1)
            else:
                y_k_j_hat = torch.zeros(num_queries, device=device)

            # Combined cost
            cost_yk = torch.abs(y_k_j_hat[:, None] - y_k_i[None, :])
            C = (self.cost_point * cost_point + 
                 cost_yk + 
                 self.cost_class * (1 - C_hat_j[:, None]))

            # Solve assignment
            C = C.cpu()
            indices_b = linear_sum_assignment(C.numpy())
            indices.append((
                torch.as_tensor(indices_b[0], dtype=torch.int64, device=device),
                torch.as_tensor(indices_b[1], dtype=torch.int64, device=device)
            ))

        return indices
    
def build_matcher_crowd(args):
    return KMOHungarianMatcher_Crowd(cost_class=args["set_cost_class"], cost_point=args["set_cost_point"], k_nearest_neighbors=3)`

@dk-liang share your suggestions too
